pkg/client/cache/fifo.go

```diff
@@ -22,12 +22,15 @@ import (
 	"k8s.io/kubernetes/pkg/util/sets"
 )
 
+type PopProcessFunc func(interface{}) error
+
 // Queue is exactly like a Store, but has a Pop() method too.
 type Queue interface {
 	Store
 
-	// Pop blocks until it has something to return.
-	Pop() interface{}
+	// Pop blocks until it has something to process.
+	// It returns the object that was process and the result of processing.
+	Pop(PopProcessFunc) (interface{}, error)
```

Comment:
I definitely agree that it's not an ideal solution. But fixing it properly can't be done within an hour or - it would require days, which we don't have before 1.3. And this is fixing the problem that we are facing now.
And I completely agree that fixing it properly after 1.3 is a good thing to do.